model/src/test/kotlin/utils/LicenseResolverTest.kt

```diff
@@ -266,5 +267,32 @@ class LicenseResolverTest : WordSpec() {
                 result should containExactlyInAnyOrder("BSD-2-Clause", "BSD-3-Clause", "Apache-2.0", "MIT")
             }
         }
+
+        "Given a package without scan result and another package with scan result, getDetectedLicensesForId()" should {
```

Comment:
This continues the pattern followed in this file of using a quite specific "outer" description, followed by only a single test case inside the `should` clause. Personally, I'd prefer to see `should` clause being used to group similar test cases, like we do elsewhere. This could be archived by a follow-up commit that rearranges all currently existing test cases in this test like

        "getDetectedLicensesForId()" should {
            "return all detected licenses for a package with a scan result for VCS" { /* ... */ }
            "return all detected licenses for a package with a scan result for source artifact" { /* ... */ }
            // ...
        }

which IMO also better highlights the differences between the test cases. What do you think?

Reply:
Roughly speaking the reason why I arrived writing this variant is because I wanted to avoid `testing too many things at a time` and I wanted the test case name to clearly describe `what` the test is about. Imo these are two common problems in `ORTs` code base. These are e.g. test case structured like "...should return the correct result", "should return the expected X", ...you name it. 

To address the above I thought I'd just name the tests following the "Given When Then" structure. When I tried this out in Kotlin (1.5 years ago) I first picked `StringSpec` which I dropped as it uses single strings which became too long. In order to work around this I just (maybe mis-)used the `ShouldSpec` which gives two lines of strings for the test case name. Obviously if the first string (the one before the should) is just the function name, this work around is less effective, e.g. the limit to test case names becomes shorter (assuming multiline strings are not an option).

Regarding what we do with the code in this file I don't have any strong opinion. I'd be fine with converting it as you suggested as long as I have a bit of freedom for defining the `inner string part` of the test case name. From [this maybe not very good example](https://github.com/kotest/kotest/blob/master/doc/styles.md#should-spec) the `ShouldSpec` seems to really be used to group tests by `function under test name`. But there are other specs available I assume we are free to use in ORT as well, like the [BevhaiorSpec](https://github.com/kotest/kotest/blob/master/doc/styles.md#behavior-spec). 

Regarding conventions for the whole project I'd personally make some more research before making any call, and further I think we should agree before as a team to write down conventions how we want to write tests. What do you think?

Reply:
> Roughly speaking the reason why I arrived writing this variant is because I wanted to avoid `testing too many things at a time`

I'm not sure I understand how your proposal helps to avoid testing too many things at a time, as the code in the inner-most function actually stays the same with both approaches.

> To address the above I thought I'd just name the tests following the "Given When Then" structure.

Frankly, I've always disliked the "Given When Then" structure. That's mostly because in my view that structure feels quite unnatural, somehow "reverse" and verbose, but at the same time with often not being clear (to me) what should go into "Given" and what into "When" (though I'm aware of the formal definitions).

That was actually one of the reasons why I decided against e.g. [using Spek as a BDD test framework](https://www.spekframework.org/gherkin/) in favor of the more flexible *kotlintest* / *kotest*.

> But there are other specs available I assume we are free to use in ORT as well, like the [BevhaiorSpec](https://github.com/kotest/kotest/blob/master/doc/styles.md#behavior-spec).

Yes, indeed, *if* you decide to use BDD-style tests, please pick the suitable spec instead of trying to apply a spec that is meant for another purpose.

> Regarding conventions for the whole project I'd personally make some more research before making any call, and further I think we should agree before as a team to write down conventions how we want to write tests. What do you think?

I was hoping to be able to mostly avoid the overhead of documenting the style of the current code base in a conventions document, but looks like this becomes necessary now. Probably as part of [CONTRIBUTING.md](https://github.com/oss-review-toolkit/ort/blob/master/CONTRIBUTING.md#coding-conventions).

> edit: could you give me a hint on what value grouping by function name (by means of using should) brings?

I value with that approach that you can quickly see which functions have test coverage, and roughly to what extent (by the number of test cases). And as test specs grow, you can quickly collapse tests for functions you're currently not interested in to get a better overview.

Reply:
> I'm not sure I understand how your proposal helps to avoid testing too many things at a time, as the code in the inner-most function actually stays the same with both approaches.

I meant that for a test which test too many things at a time it would be impossible to define a good 'given when then' test case name and thus the code in the inner-most function would need to change in order to arrive at a good test case name. 

> I was hoping to be able to mostly avoid the overhead of documenting the style of the current code base in a conventions document, but looks like this becomes necessary now. Probably as part of CONTRIBUTING.md.

I didn't want to push us into writing this now now. As long as I still have the freedom to use the mentioned BDD spec I'm still fine with the status quo.